ROFO-42 회원가입 API 작성

Unagi-zoso opened the pull request:
### 개요

### 변경사항
- User Entity 프로퍼티 추가, flyway script 작성
- security 패키지 이동 user -> auth
- ImageService 클래스 작성
- KakaoAPI 호출기능 작성
- SwaggerConfig Authorization 헤더 시큐리티 스키마 추가
- 회원가입 관련 기능 및 테스트 작성
- 회원가입 API 문서 작성
- WebPValidator 작성
- Controller 단 통합테스트 어노테이션 작성

**피드백 반영**
- MaxFIleSize Validator 작성
- 중복닉네임 검사 API, 로직 작성

### 테스트
- 테스트 코드 추가여부 O

### 관련 지라 및 위키 링크
 - [JIRA](https://weit-2nd.atlassian.net/browse/ROFO-42) 


### 리뷰어에게 하고 싶은 말
어떤 피드백이라도 감사히 잘 받겠습니다!
Swagger 로 문서를 작성하며 DTO 에 스키마 정보를 남기기 싫어
Spec 에 SchemaProperty 필요한 부분들을 정의했습니다.
그냥 하드코딩 느낌으로 하나하나 다 정의하다보니 이후에 변경할 때 번거로울 수 있을 것 같았지만 
그래도 DTO 를 더럽히지 않는다는 점과 MultipartFile 의 경우 @Schema 를 설정해도 적용이 안되는 문제가 있어 현재 방식을 고수하기로 했습니다.
현재 API 문서에서 테스트를 위해선 실제 Kakao Access Token 이 필요합니다. 
전 이 부분을 위해 개인적으로 Kakao Acces Token 을 발급 받고 테스트를 진행했습니다. 
좀 번거로운 부분이 있었는데 이 부분을 개선할 방법이 있을지 고민하고 있습니다.

Unagi-zoso committed:
ROFO-42 User Entity socialId, profileImageName 추가 및 반영

Unagi-zoso committed:
ROFO-42 UserRepository socialId 관련 메소드, 테소트 코드 추가

Unagi-zoso committed:
ROFO-42 security 패키지 이동 user -> auth

Unagi-zoso committed:
ROFO-42 ImageService 클래스, 테스트 코드 작성

Unagi-zoso committed:
ROFO-42 Auth Test Fixture 및 테스트 이미지 데이터 추가

Unagi-zoso committed:
ROFO-42 Kakao 사용자 정보 요청 메소드 작성 및 테스트

Unagi-zoso committed:
ROFO-42 AuthCommandService 및 회원가입 DTO, 커스텀 예외 생성

Unagi-zoso committed:
ROFO-42 AuthIntegrationServiceTest 작성

Unagi-zoso committed:
ROFO-42 Swagger 태그 추가 및 Authorize 헤더 추가

Unagi-zoso committed:
ROFO-42 Tika 종속성 추가 및 WebPValidator 추가

Unagi-zoso committed:
ROFO-42 기존 커스텀 예외 메시지 가시성 변경 private -> public

Spec 에서도 사용할 수 있게 했습니다.

Unagi-zoso committed:
ROFO-42 AuthController 클래스, 테스트, spec 작성

Unagi-zoso committed:
ROFO-42 DTO 프로퍼티명 변경 및 스웨거 어노테이션 제거

Unagi-zoso committed:
ROFO-42 generateSocialId() 메소드 명 변경 -> obtainSocialId()

해당 메소드는 외부 API 요청하는 부분을 포함한다. 그런 의미를 담을 수 있는 메소드명을 고민하다 바꿨다.

Unagi-zoso committed:
ROFO-42 AuthIntegrationControllerTest 추가

SignUpRequest 바인딩 이슈 문제를 잡기 위해 컨트롤러단 통합테스트 코드를 추가했다.

Unagi-zoso committed:
ROFO-42 사용하지 않는 어노테이션 제거

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/application/service/AuthCommandService.kt:
```diff
@@ -0,0 +1,57 @@
+            userRepository.save(user)
+            userAgreedTermCommandService.storeUserAgreedTerms(user, signUpRequest.agreedTermIds)
+            imageService.upload(imageName, signUpRequest.profileImage) // 외부 저장소 롤백 어려움을 고려해 가장 아래에 배치
+        }
```
아래와 같이 공통 부분을 아래와 같이 묶어서 밖으로 빼내면 어떨까요??
```kotlin
        val user = User.of(socialId, signUpRequest.nickname)
        userRepository.save(user)
        userAgreedTermCommandService.storeUserAgreedTerms(user, signUpRequest.agreedTermIds)

        if (signUpRequest.profileImage != null) {
            val imageName = imageService.generateImageName(signUpRequest.profileImage.originalFilename)
            user.profile = Profile(signUpRequest.nickname, imageName)
            imageService.upload(imageName, signUpRequest.profileImage) // 외부 저장소 롤백 어려움을 고려해 가장 아래에 배치
        }
```

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/dto/SignUpRequest.kt:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.web.multipart.MultipartFile
+
+data class SignUpRequest(
+    @field:Pattern(regexp = NICKNAME_REGEX_STR, message = NICKNAME_REGEX_DESC)
+    val nickname: String = "",
+    @field:WebPImage
+    val profileImage: MultipartFile?,
+    val agreedTermIds: Set<Long> = emptySet(),
+    val socialLoginType: SocialLoginType = SocialLoginType.KAKAO,
```
여기에 있는 필드들에 기본값이 있는 이유를 알수있을까요??

Unagi-zoso replied:
설계한 API 의도대로 요청하지 않는 경우 (ex nickname 폼을 누락)
nickname 은 null 을 허락하지 않는 자료형이라 비즈니스와 관련 없는 예외가 발생하게 됩니다. 따라서 nickname 폼을 누락하더라도 기본적으로 nickname 이 생성되도록 두었습니다. (이후 유효성 검사에서 예외가 발생합니다.)

agreedTermIds 도 비슷한 맥락 입니다. (register() 내 약관 검사에서 예외가 발생합니다.)

socialLoginType 의 경우 현재 카카오 로그인 만을 허용하기에 외부에서 입력을 받지 않습니다. 따라서 기본값을 설정해두었습니다.

konempty replied:
> nickname 은 null 을 허락하지 않는 자료형이라 비즈니스와 관련 없는 예외가 발생하게 됩니다.

제 기억에는 HttpNotReadable 대충 요런 이름의 에러가 발생했던것 같은데
그 에러를 잡아서 ${변수명}은 null일수 없습니다
라는 에러를 내리면 비지니스와 관련있는 에러로 처리가능하지 않을까요??

konempty replied:
조금 찾아보니까 제 기억과 비슷한 코드가 잇네요

https://yeongchan1228.tistory.com/m/133

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/api/AuthController.kt:
```diff
@@ -0,0 +1,33 @@
+    @PostMapping(consumes = [MULTIPART_FORM_DATA_VALUE])
+    override fun signUp(
+        @RequestHeader(AUTHORIZATION) socialAccessToken: String?,
+        @Valid @ModelAttribute signUpRequest: SignUpRequest,
```
ModelAttribute를 붙히는 이유가 뭘까요..?

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/api/AuthController.kt:
```diff
@@ -0,0 +1,33 @@
+    ) {
+        socialAccessToken?.let {
+            authCommandService.register(SocialAccessToken(socialAccessToken), signUpRequest)
+        } ?: throw IllegalArgumentException("socialAccessToken 가 존재하지 않습니다.")
```
이 코드도 코틀린의 문법을 충분히 활용한것 같긴한데 좀더 적극적으로 활용하면 같은코드를 좀더 짧게 아래와 같이 가능합니다!
```kotlin
requireNotNull(socialAccessToken) { "socialAccessToken 가 존재하지 않습니다." }
authCommandService.register(SocialAccessToken(socialAccessToken), signUpRequest)

// 혹은 합쳐서
authCommandService.register(
    SocialAccessToken(requireNotNull(socialAccessToken) { "socialAccessToken 가 존재하지 않습니다." }),
    signUpRequest,
)
```

lin-chae commented on src/main/kotlin/kr/weit/roadyfoody/auth/exception/UserAlreadyExistsException.kt:
```diff
@@ -0,0 +1,9 @@
+import kr.weit.roadyfoody.common.exception.ErrorCode
+
+class UserAlreadyExistsException : BaseException(
+    ErrorCode.INVALID_REQUEST,
```
리소스가 이미 존재하는 경우 “ 새로운 리소스 추가” 작업의 실패를 표현을 위해 409 에러도 괜찮지 않을까 싶습니다!

konempty replied:
좋은 의견이네요 👍

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/spec/AuthControllerSpec.kt:
```diff
@@ -0,0 +1,146 @@
+                                        example = "[1, 2, 3]",
+                                    ),
+                            ),
+                        ],
```
우와 ㅋㅋㅋㅋ 상상만 했던 코드인데 눈으로 직접보니까 재미있네요 ㅋㅋㅋㅋ
근데 호불호 갈리기 딱좋은 코드일것 같은데 저는 이정도면 아슬아슬하게 호네요

lin-chae reviewed (approved):
굳굳! 파일에 대한 처리를 잘해주셨네요! 고생하셨습니담! 👍

konempty commented on src/main/kotlin/kr/weit/roadyfoody/global/service/ImageService.kt:
```diff
@@ -0,0 +1,28 @@
+        storageService.upload(name, file.inputStream)
+    }
+
+    fun generateImageName(fileName: String?): String {
```
파일이름이 null일 경우가 있을까요??

Unagi-zoso replied:
정상적인 서비스 환경에선 문제가 없을듯 하네요.
테스트 환경에서 MockMultipartFile 는 originalFilename 이 Nullable 해서 저렇게 두었는데 StandardMultipartFile 에선 문제 없는것으로보입니다. 
이 부분은 고치겠습니다

konempty commented on src/main/kotlin/kr/weit/roadyfoody/global/service/ImageService.kt:
```diff
@@ -0,0 +1,28 @@
+            if (fileName == null) {
+                ""
+            } else {
+                ".${StringUtils.getFilenameExtension(fileName)}"
```
요런 코드가 예전에 저희 부서에서도 이슈가 된적이 있었는데
유저가 jpeg파일을 png라고 파일 확장자만 바꿔서 업로드 한적이 있는데
그 파일이 대부분 문제 없었지만 일부 서비스에서 읽지를 못하는 경우가 있었어요..!
그래서 파일의 이름에서 확장자를 고대로 따오는게 별로 안전하지는 않은 코드라고 생각이 드네요 ㅋㅋㅋ큐ㅠㅠ

konempty commented on src/main/kotlin/kr/weit/roadyfoody/global/validator/WebPValidator.kt:
```diff
@@ -0,0 +1,25 @@
+    override fun isValid(
+        p0: MultipartFile?,
+        p1: ConstraintValidatorContext?,
+    ): Boolean = p0 == null || Tika().detect(p0.inputStream) == "image/webp"
```
Tika 객체를 매번 생성하는게 아닌 한번 생성해둔 객체를 재활용하는게 어떨까요??
참고로 Tika().detect는 threadSafe한 메소드라고하네요 ㅋㅋㅋ
https://www.mail-archive.com/user@tika.apache.org/msg03473.html

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthCommandServiceTest.kt:
```diff
@@ -0,0 +1,74 @@
+                verify(exactly = 0) { imageService.generateImageName(any<String>()) }
+                verify(exactly = 1) { userRepository.save(any<User>()) }
+                verify(exactly = 1) { userAgreedTermCommandService.storeUserAgreedTerms(any<User>(), any<Set<Long>>()) }
+                verify(exactly = 0) { imageService.upload(any<String>(), any<MultipartFile>()) }
```
호출 횟수가 겹치는 애들이 있는데 묶어주면 어떨까요??
```kotlin
                    verify(exactly = 1) {
                        authQueryService.requestKakaoUserInfo(any<SocialAccessToken>())
                        userRepository.existsBySocialId(any<String>())
                        termCommandService.checkRequiredTermsOrThrow(any<Set<Long>>())
                        userRepository.save(any<User>())
                        userAgreedTermCommandService.storeUserAgreedTerms(any<User>(), any<Set<Long>>())
                    }
                    verify(exactly = 0) {
                        imageService.generateImageName(any<String>())
                        imageService.upload(any<String>(), any<MultipartFile>())
                    }
```

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthCommandServiceTest.kt:
```diff
@@ -0,0 +1,74 @@
+            every { userAgreedTermCommandService.storeUserAgreedTerms(any<User>(), any<Set<Long>>()) } just runs
+            every { imageService.upload(any<String>(), any<MultipartFile>()) } just runs
+            then("이미지를 업로드하고 User 가 생성된다.") {
+                authCommandService.register(TEST_SOCIAL_ACCESS_TOKEN, createTestSignUpRequest())
+                verify(exactly = 1) { authQueryService.requestKakaoUserInfo(any<SocialAccessToken>()) }
+                verify(exactly = 1) { userRepository.existsBySocialId(any<String>()) }
+                verify(exactly = 1) { termCommandService.checkRequiredTermsOrThrow(any<Set<Long>>()) }
+                verify(exactly = 1) { imageService.generateImageName(any<String>()) }
+                verify(exactly = 1) { userRepository.save(any<User>()) }
+                verify(exactly = 1) { userAgreedTermCommandService.storeUserAgreedTerms(any<User>(), any<Set<Long>>()) }
+                verify(exactly = 1) { imageService.upload(any<String>(), any<MultipartFile>()) }
```
여긴 다묶을수 있겠네요 ㅋㅋㅋ

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthIntegrationServiceTest.kt:
```diff
@@ -0,0 +1,131 @@
+            userRepository.findAll()
+                .forEach {
+                    it.profile.profileImageName?.let { imageName ->
+                        s3Template.deleteObject(s3Properties.bucket, imageName)
```
ㅋㅋㅋㅋ 저번에 테스트 짤때 주석으로 써놨는데
S3의 스펙상 파일삭제는 비동기로 이루어지기 때문에 삭제 요청을 보내도 실제 삭제는 언제될지 알수 없어서
테스트에서는 지우는게 큰 의미가 없어요

Unagi-zoso replied:
아하 그렇군요!
일단 저 부분을 제거하면 bucket 비어있지 않다고 deleteBucket() 이 실패합니다.
저 부분이 있더라도 deleteBucket() 이 실행되는 시점에 deleteObject() 가 완료되지 않아 또 deleteBucket() 이 실패할 수도 있다는 말씀이신가요?

konempty replied:
아하 deleteBucket을 위한 코드라면 있어야 합니다!
저도 제 코드에 그것때문에 삭제코드 넣어놓고 그렇게 해야 한다는걸 잊고있었네요 ㅋㅋㅋㅋ

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthIntegrationServiceTest.kt:
```diff
@@ -0,0 +1,131 @@
+        }
+
+        val authQueryService: AuthQueryService = mockk<AuthQueryService>()
+        val authCommandService =
```
요건 코멘트는 아니고 그냥 지식 공유인데
테스트 대상 객체는 다른 객체와는 다르게 sut(System under test)이라는 
정해진 이름으로 지어두는걸 선호하는 회사들도 많아요!
https://en.wikipedia.org/wiki/System_under_test

jifrozen0110 commented on src/test/kotlin/kr/weit/roadyfoody/auth/presentation/api/AuthIntegrationControllerTest.kt:
```diff
@@ -0,0 +1,92 @@
+@Profile("test")
+@Primary
+@Component
+class MockKakaoClient : KakaoClientInterface {
```
테스트를 위한 mock 구현체 만드는거 좋네요!

konempty replied:
<img width="937" alt="스크린샷 2024-06-23 오후 6 42 41" src="https://github.com/Weit-2nd/roady-foody-server/assets/44064530/24101ad2-cc5a-40b3-94bd-80e1e7301319">
요렇게 mockBean으로 처리하는게 어떨까요??

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/application/service/AuthQueryService.kt:
```diff
@@ -0,0 +1,12 @@
+class AuthQueryService(
+    private val kakaoClientInterface: KakaoClientInterface,
+) {
+    fun requestKakaoUserInfo(socialAccessToken: SocialAccessToken) = kakaoClientInterface.requestUserInfo(socialAccessToken)
```
요기에서 에러가 발생했을때 에러처리 하는 부분이 없는것 같은데
요 API가 실패한다면 잘못된 토큰을 보냈다는 에러 메시지가 필요하지 않을까요??
안그러면 에러 발생시 500 에러가 내려가지 않을까 싶네요 ㅋㅋ큐ㅠㅠ

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthQueryServiceTest.kt:
```diff
@@ -0,0 +1,31 @@
+        val authQueryService = AuthQueryService(kakaoClientInterface)
+
+        given("requestKakaoUserInfo 테스트") {
+            `when`("카카오 Access Token 을 가져오면") {
```
여기도 실패하는 경우에 대한 에러 처리도 있으면 어떨까 싶네요! ㅋㅋㅋ

jifrozen0110 commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthIntegrationServiceTest.kt:
```diff
@@ -0,0 +1,131 @@
+
+        given("프로필 사진이 존재하는 경우") {
+            `when`("회원가입을 요청하면") {
+                every { authQueryService.requestKakaoUserInfo(any<SocialAccessToken>()) } returns createTestKakaoUserResponse()
```
모든 given마다 존재하는데
beforeEach로 묶는거 어떨까요?

konempty commented on src/test/kotlin/kr/weit/roadyfoody/global/service/ImageServiceTest.kt:
```diff
@@ -0,0 +1,44 @@
+        `when`("확장자가 없는 이미지 이름을 전달하면") {
+            then("확장자가 없는 이미지 이름이 생성된다.") {
+                val imageName = imageService.generateImageName(TEST_WEBP_IMAGE.originalFilename.removeSuffix(TEST_WEBP_EXTENSION))
+                StringUtils.getFilenameExtension(imageName) shouldBe "null"
```
StringUtils은 저희가 개발한 코드가 아닌데 저희가 테스트 하는게 뭔가 어색하네요 ㅋㅋㅋ

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/presentation/api/AuthIntegrationControllerTest.kt:
```diff
@@ -0,0 +1,92 @@
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
+
+@ControllerIntegrateTest
+class AuthIntegrationControllerTest(
```
AuthControllerTest와 AuthIntegrationControllerTest로 나눠서 테스트된 이유가 뭘까요??
둘다 유닛 테스트로 짜거나 둘다 통합테스트로 통일을 하는게 낫지 않을까요?

Unagi-zoso replied:
일단 컨트롤러 테스트 만으로도 대부분의 상황은 테스트가 가능했습니다. (데이터 바인딩, 유효성 검사..). 
이후에 sandbox 에 배포해서 테스트를 해보니 SignUpRequest 객체에 agreedTermIds 가 제대로 전달이 안됐고 이를 판별하는 기능은 service 단에 구현되어 있어서 서비스를 전부 모킹한 기존 컨트롤러 테스트에서 이런 문제를 테스트하지 못했습니다.
지금와서는 기존 컨트롤러 테스트에서 패러미터 바인딩 체크 부분을 추가하면 됐던게 아닌가 생각이 듭니다만 슬슬 프로젝트에서 통합테스트 관련한 코드가 보여 이참에 이런식으로 해결해보면 좋지 않을까 생각이 들어 적어보았습니다.

konempty replied:
으흠 이슈가 잘이해안되는데 일단 제가 이해한대로 답변을 드리자면
서비스에서 필수여부를 판단하고 있었다면 그건 컨트롤러의 역할이 아니라 서비스의 역할로 잡으신것 같은데
그렇다면 컨트롤러 테스트에서 확인할게 아니라 서비스 테스트에서 확인해야하는거 아닐까요..??

Unagi-zoso replied:
그렇네요. 제가 좀 오용한 듯 하네요. 
저의 원래 의도는 SignUpRequest 객체를 제대로 바인딩 하는지를 확인하기 위해서 였는데 이를 위해 통합테스트를 할 필요는 없어보이네요. 
이 부분은 제거하고 다른 부분에서 보완하겠습니다.

jifrozen0110 commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthIntegrationServiceTest.kt:
```diff
@@ -0,0 +1,131 @@
+                imageService,
+            )
+
+        given("프로필 사진이 존재하는 경우") {
```
개인적인 선호도 문제인것같은데
저는 given은 맥락을 명시하고
when으로 특정한 상황을 명시하는걸 선호해서
저라면 AuthCommandServiceTest처럼
given으로 회원가입하는 경우를 묶고
when으로 프로일이 없는경우 있는 경우...
이렇게 나눴을 것 같습니다.

jifrozen0110 commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthQueryServiceTest.kt:
```diff
@@ -0,0 +1,31 @@
+
+class AuthQueryServiceTest :
+    BehaviorSpec({
+        val kakaoClientInterface = mockk<KakaoClientInterface>()
```
MockKakaoClient를 생성하였는데 여기선 mockk을 사용해 mock화한 이유가 있을까요?

Unagi-zoso replied:
좋은 질문 감사합니다.
AuthQueryService 를 테스트 하기 위한 코드라 별도의 빈을 가져오기 보단 모킹을 통해서 일관성있게 구성하면 좋을 것 같아서 그랬습니다.
MockKakaoClient 를 만든 이유는 컨트롤러 단 **통합 테스트** 시 실제 API 를 호출하는 비용이 부담스러워 그 부분을 대체하기 위함이였습니다.

jifrozen0110 replied:
아하 알겠습니다!

jifrozen0110 commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthIntegrationServiceTest.kt:
```diff
@@ -0,0 +1,131 @@
+                    }
+                }
+        }
+        afterTest { clearAllMocks() }
```
clearAllMocks을 따로 명시한 이유가 궁금합니다!
mockk에서 자동으로 mock 객체들을 초기화해주지 않나요?

Unagi-zoso replied:
일단 제 코드에서 해당 부분을 제거하면 테스트에서 실패합니다.

제가 파악하고 있는 것으론 현재 코드는 Spec 내부에서 변수들을 선언했습니다.
제가 알기로 Kotest 의 Spec 단 한 번만 생성되어(기본설정) Spec 내 테스트에서 사용되는 것으로 알고 있습니다.  따라서 한 번 생성된 mock 객체를 반복해서 사용하고 verify(exactly = 1) 을 하면 실패하게 됩니다. (초기화 되지 않아 중첩되기 때문)

mockk 자동 초기화는 제가 잘 모르는 부분이라 좀 더 공부해보겠습니다!

jifrozen0110 replied:
오호 그렇군요 저도 잘 몰라서 여쭤봤습니다. 감사합니다.

jifrozen0110 reviewed (approved):
고생하셨습니다.
통합테스트와 단위테스트를 나눠 꼼꼼하게 테스트를 작성해주셨네요. 배워갑니다!

konempty commented on src/test/kotlin/kr/weit/roadyfoody/user/fixture/UserFixture.kt:
```diff
@@ -8,9 +10,30 @@ val TEST_MIN_LENGTH_NICKNAME = (0 until NICKNAME_MIN_LENGTH).map { 'a' + it }.jo
+const val TEST_USER_PROFILE_IMAGE_NAME = "test_image_name"
 
-fun createTestUser(id: Long) = User(id, "$TEST_USER_NICKNAME$id")
+fun createTestUser(id: Long = TEST_USER_ID) =
+    User(
+        id,
+        TEST_USER_SOCIAL_ID,
+        Profile(
+            "$TEST_USER_NICKNAME$id",
+            "${TEST_USER_PROFILE_IMAGE_NAME}_$id",
+        ),
+    )
+
+fun createTestUser(
+    nickname: String = TEST_USER_NICKNAME,
+    socialId: String = TEST_USER_SOCIAL_ID,
+) = User.of(
+    socialId,
+    nickname,
+    TEST_USER_PROFILE_IMAGE_NAME,
+)
```
요 두개의 메소드를 하나로 합칠수 있을것 같은데 함치면 어떨까요??

konempty commented on src/test/resources/image/gif_test_img.gif:
ㅋㅋㅋㅋ 요렇게 테스트용 리소스는 올리지 않고 테스트 할때 동적으로 생성하는게 좋아요!
gif는 좀 어렵지만 jpg나 png는 가능할거예요!

Unagi-zoso replied:
혹시 이유를 알 수 있을까요?
정적 데이터 IO 비용을 의식해서 인가요??

konempty replied:
저렇게 테스트를 위한 리소스를 추가하다보면 프로젝트의 크기가 프로덕션 코드와 리소스로 인해 커지는게 아니라 테스트 리소스로 인해 커질것 같아서 말씀드렸습니다~!

konempty reviewed (changes requested):
고생하셨습니다 👍 👍 
회원가입 API가 빠르게 나왔군요! ㅋㅋㅋㅋ
코멘트 조금 남겼으니 조금만더 수정 부탁드립니다~!

Unagi-zoso committed:
ROFO-42 사용하지 않는 코드 및 자료 제거

Unagi-zoso committed:
ROFO-42 WebP 파일 변환 라이브러리 추가

손쉽게 WebP 파일로 변환할 수 있어 추가했습니다. 어느 라이브러리를 쓸게 고민하다.
최근까지 관리되고 개발자분이 Kotest 를 운영하는 등 경험적으로도 믿을만하다 생각해 채택했습니다.

[Maven Repo](https://mvnrepository.com/artifact/com.sksamuel.scrimage/scrimage-webp)
[Github](https://github.com/sksamuel/scrimage/issues)

Unagi-zoso committed:
ROFO-42 ImageBytesGenerator 및 ImageFormat Enum class 작성

Unagi-zoso committed:
ROFO-42 MimeUtils 작성 (Tika 재사용 클래스)

Unagi-zoso committed:
ROFO-42 Test Fixture 피드백 반영

Unagi-zoso committed:
ROFO-42 WebPValidator 클래스에 MimeUtil 적용

Unagi-zoso committed:
ROFO-42 ObjectMapperProvider 소속 패키지 변경

기존엔 자신만 쓰일 것 같은 jsonmapper 패키지에 들어있었다. MimeUtil 이 생기고
둘의 유틸리티적 성향이 비슷하다 생각해 utils 라는 패키지에서 관리하도록 두었다.

Unagi-zoso committed:
ROFO-42 generateImageName 메소드 MimeUtils 적용

기존 방식은 파일 이름에서 확장자를 유추했는데 이는 문제를 일으킬 수 있다.

Unagi-zoso committed:
ROFO-42 카카오 사용자 정보 요청 API 예외발생 시 처리 부분 작성

기본적으로 401(UnAuthorized) 를 상태번호로 반환합니다.

Unagi-zoso committed:
ROFO-42 AuthController Null check 로직 변경 및 테스트 코드에 Enum class 적용

Unagi-zoso committed:
ROFO-42 SignUpRequest 자료형 Nullable 하게 변경

기존 방식은 코틀린에서 Null 을 허용하지 않아 스프링의 multipart 로 오브젝트를 만들 시 ObjectError 가 발생했다.
이는 현재 우리 프로젝트의 전역 예외 처리에서는 다루지 않으며 다룬다해도 예외 메시지가 난해하고 서비스 패키지 내용을 알 수 있어
클라이언트로 넘겨주기 애매했다. 따라서 Nullable 하게 두며 Jakarta 유효성 검사로 Nullcheck 를 했다. 이때 null 이 들어온다면
이번엔 FieldError 로 발생해 우리 예외 처리 시스템에서 잘 처리할 수 있다.

대신 request 오브젝트를 사용하는 부분에선 Null 이 아님을 처리해줘야한다. !! 를 사용해서 번거로움을 해소해도 될까 고민이다.

Unagi-zoso committed:
ROFO-42 중복 회원가입 시도 시 반환 상태번호 변경

Unagi-zoso committed:
ROFO-42 테스트 코드 lifecycle hook 변경

각 테스트 상황 시에만 적용되어도 괜찮을 것 같아 변경했습니다.

Unagi-zoso committed:
ROFO-42 SignUp 문서 최신화

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/dto/SignUpRequest.kt:
```diff
@@ -0,0 +1,21 @@
+import org.springframework.web.multipart.MultipartFile
+
+data class SignUpRequest(
+    @field:NotNull(message = NOT_NULL_MSG)
+    @field:Pattern(regexp = NICKNAME_REGEX_STR, message = NICKNAME_REGEX_DESC)
+    val nickname: String?,
+    @field:WebPImage
+    val profileImage: MultipartFile?,
+    @field:NotNull(message = NOT_NULL_MSG)
+    val agreedTermIds: Set<Long>?,
+    val socialLoginType: SocialLoginType = SocialLoginType.KAKAO,
```
아하 ㅠㅠ notNull을 제대로 쓰기위해 모두가 nullable이 되어버렸군요
이게 이전 코드보다 더 좋지 못한 코드가 되어버린것 같아서 슬프네요

Unagi-zoso replied:
저도 이걸 좀 고민했는데 기본값을 주지 않고 null 을 허용하지 않는 자료형을 사용할 시
스프링에서 바인딩을 할 때 notnull 자료형에 null 을 넣었다고 예외가 발생하더라구요.
이게 fieldError 라면 저희 중앙예외처리 방식으로 처리가 가능한데 이 경우엔objectError 로 예외가 발생하더라구요.

```
// ExceptionHandler.kt 일부

    private fun MethodArgumentNotValidException.messages(): List<String> {
        return bindingResult.fieldErrors.map { "${it.field}: ${it.defaultMessage.orEmpty()}" }
    }

```

그래서 globalError(ObjectError 콜렉션) 으로 예외 정보를 얻어 반환하면 안될까 싶었는데 objectError 의 경우 어느 필드가 null 이다 라고 추출하기가 힘든 완벽한 영어 문장이 나와 이걸 쓰기가 힘들어보였습니다. 

그래서 null 을 허용하는대신 jakarta 유효성 검사로 필드를 검사하면 fieldError 로 나와
정상적으로 저희 중앙예외처리 방식을 사용할 수 있어서 이 방식을 택하긴 했는데

솔직히 이 방식을 사용하면서 회의감이 많이 들었네요.. 개선할 방법이 있을까요?

개인적으론 기본값을 두던 처음 방식은 어땠을까 하는 생각이 좀 드네요..

Unagi-zoso replied:
https://yeongchan1228.tistory.com/m/133

이 사례를 알려주셨는데 이건 Json 데이터의 경우에 해당되는 예외로 보였습니다.
저의 경우엔 Multipart/form-data 를 사용하고 있어 동적으로 생성자를 찾고 인스턴스를 생성하며 `MethodArgumentNotValidException` 예외가 발생했었네요.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/application/service/AuthQueryService.kt:
```diff
@@ -0,0 +1,19 @@
+    private val kakaoClientInterface: KakaoClientInterface,
+) {
+    fun requestKakaoUserInfo(socialAccessToken: SocialAccessToken): KakaoUserResponse =
+        runCatching {
+            kakaoClientInterface.requestUserInfo(socialAccessToken)
+        }.getOrElse {
+            throw InvalidTokenException()
+        }
```
제 처음 의도는 에러 코드에 따른 에러의 분기였는데
모든 에러를 잡아서 InvalidTokenException으로 내리는군요
일단은 요렇게 가시죠

Unagi-zoso replied:
저도 그 부분을 좀 고민을 했었네요.

```
// RestClientConfig.kt 일부
  restClientBuilder
            .requestInterceptor(RetryClientHttpRequestInterceptor())
            .requestFactory(clientHttpRequestFactory())
            .defaultStatusHandler(HttpStatusCode::is4xxClientError) { _, response ->
                log.error("Client Error Code={}", response.statusCode)
                log.error("Client Error Message={}", String(response.body.readAllBytes()))
                throw RestClientException()
            }
            .defaultStatusHandler(HttpStatusCode::is5xxServerError) { _, response ->
                log.error("Server Error Code={}", response.statusCode)
                log.error("Server Error Message={}", String(response.body.readAllBytes()))
                throw RestClientException()
            }
            .build()

```

제가 요청을 해서 400, 500 번대의 응답을 받아도 이 핸들러 부분에서 재킹을 하는 것 같더라구요.

RestClientException 예외에서 실패 시의 정보를 구할 수도 없어 이 부분을 어떻게 동적으로 잘 처리할 수 있을까 고민하다. 대부분의 경우엔 토큰이 유효하지 않을 때가 아닐까 싶어 일단 이렇게 처리해뒀습니다.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/api/AuthController.kt:
```diff
@@ -0,0 +1,31 @@
+    @PostMapping(consumes = [MULTIPART_FORM_DATA_VALUE])
+    override fun signUp(
+        @RequestHeader(AUTHORIZATION) socialAccessToken: String?,
+        @Valid signUpRequest: SignUpRequest,
```
@ RequestBody를 붙혀야 하지 않을까요..??

Unagi-zoso replied:
혹시 제가 multipart/form-data 로 받을 예정인데 이 경우 json 역직렬화 같은 과정이 필요없어 @RequestBody 는 없어도 되는 것 아닌가요..?

konempty replied:
아 멀티 파트니까 requestPart겠네요 ㅋㅋㅋ 
제 의도는 아래의 링크처럼 처리하는걸 생각했는데 안붙혀도 잘 작동하나요??
https://github.com/weIT-1st/Odya-server/blob/4bcdb173573c53ffb3aea2722697474f0f5ded5b/src/main/kotlin/kr/weit/odya/controller/TravelJournalController.kt#L45

Unagi-zoso replied:
작동은 잘됐습니다. 제가 원하는 대로 바인딩을 했는데 
non-null 자료형과 저희 프로젝트 중앙예외처리 시스템을 잘 이용하려면
파일과 파일외의 요소는 분리해서 다루는게 좋겠더라구요. (저의 코드에선 텍스트들은 JSON 으로 SignUpRequest, 이미지파일은 MultipartFile 로 분리해 입력 받았습니다.)
현재는 보여주신 코드도 잘 참고해서 적용해뒀습니다.

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/dto/SignUpRequest.kt:
```diff
@@ -0,0 +1,21 @@
+    @field:NotNull(message = NOT_NULL_MSG)
+    @field:Pattern(regexp = NICKNAME_REGEX_STR, message = NICKNAME_REGEX_DESC)
+    val nickname: String?,
+    @field:WebPImage
+    val profileImage: MultipartFile?,
```
<img width="283" alt="스크린샷 2024-06-25 오전 12 06 25" src="https://github.com/Weit-2nd/roady-foody-server/assets/44064530/c7b25e3e-c4a2-46fc-9e77-bc3fe0a6f83a">

사진 크기에 대한 validation이 있으면 어떨까요??

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthIntegrationServiceTest.kt:
```diff
@@ -0,0 +1,136 @@
+@Transactional
+@ServiceIntegrateTest
+class AuthIntegrationServiceTest(
+    private val s3Template: S3Template,
+    private val s3Properties: S3Properties,
+    private val termCommandService: TermCommandService,
+    private val userAgreedTermCommandService: UserAgreedTermCommandService,
+    private val userRepository: UserRepository,
+    private val imageService: ImageService,
+    private val termRepository: TermRepository,
+) : BehaviorSpec({
+        val authQueryService: AuthQueryService = mockk<AuthQueryService>()
+        val authCommandService =
+            AuthCommandService(
+                authQueryService,
+                termCommandService,
+                userAgreedTermCommandService,
+                userRepository,
+                imageService,
+            )
```
authQueryService 요거 하나를 목킹해서 AuthCommandService에 주입하기위해 불필요한 빈들을 주입받는데
그런 의도라면 아래의 코드처럼 mockkBean으로 처리하면 깔끔하게 해결됩니다~!

<img width="920" alt="스크린샷 2024-06-25 오전 12 14 29" src="https://github.com/Weit-2nd/roady-foody-server/assets/44064530/fe334305-ab0c-4b07-a70b-e2d89145d7ef">

konempty commented on src/test/kotlin/kr/weit/roadyfoody/global/utils/MimeUtilsTest.kt:
```diff
@@ -0,0 +1,80 @@
+class MimeUtilsTest :
+    BehaviorSpec({
+        given("detectMimeType 메소드") {
+            `when`("${WEBP.getStrValues()} 이미지 파일을 전달하면") {
+                then("${WEBP.getStrValues()} 이미지 MIME 타입을 반환한다.") {
```
ㅋㅋㅋㅋ... 요건 그냥 "WEBP 이미지 파일을 전달하면" 요런식으로 하드코딩 해도 괜찮았을것 같아요

Unagi-zoso replied:
아마 이런 부분을 오버엔지니어링의 요소로 말씀하신게 아닌가 생각이 드네요.
저도 코드를 작성하면서 향후에 안생길지도 모르는 부분에 대해 너무 과하게 대처하려하는게 아닌가 생각이 계속 들었네요..

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/application/service/AuthCommandServiceTest.kt:
```diff
@@ -0,0 +1,98 @@
+        beforeEach {
+            every { authQueryService.requestKakaoUserInfo(any<SocialAccessToken>()) } returns createTestKakaoUserResponse()
+            every { termCommandService.checkRequiredTermsOrThrow(any<Set<Long>>()) } just runs
+            every { imageService.generateImageName(any<MultipartFile>()) } returns UUID.randomUUID().toString()
```
imageService.generateImageName 메소드는 모킹하지 않고도 가능하지 않을까 싶은데 아래 사진처럼 imageService를 mockk이 아닌 spyk로 선언하면 어떨까요?

<img width="912" alt="스크린샷 2024-06-25 오전 12 30 23" src="https://github.com/Weit-2nd/roady-foody-server/assets/44064530/a83393aa-db7c-4ca1-b418-a5a4e92c0cb2">

konempty reviewed (changes requested):
정말 고생이 많으시네요 ㅠㅠ
하지만 코멘트 할게 조금남았는데 진짜 조금만더 수정 부탁드립니다..!

konempty commented:
그리고 한가지 안타까운건
제가 학생분들이나 취준생분들 코드에 리뷰를 하다보면 코드 정리에 너무 집중을 한 나머지
정도를 넘어선 오버엔지니어링을 하시는 분들이 있는데 주열님 코드에서 간간히 보여서 너무 안타깝네요 ㅠㅠㅠ
이게 제가 말로 설명할수 없고 갠취의 영역이라 말씀드리기 까다로운데 주열님이라면 금방 적정수준의 감을 잡으실거라 믿습니다..!

Unagi-zoso commented:
이 부분은 꼭 제가 저만의 기준을 잘 세워보겠습니다!

Unagi-zoso committed:
ROFO-42 사용되지 않는 메소드 제거 (ImageFormat.getStrValues())

Unagi-zoso committed:
ROFO-42 Auth Fixture 분리 및 multipart 전송 테스트를 위한 Fixture 추가

Unagi-zoso committed:
ROFO-42 스웨거 RequestPart 전송 시 에러 해결을 위한 빈 등록

[참고 링크](https://one-armed-boy.tistory.com/entry/Swagger-RequestPart%EB%A5%BC-%ED%86%B5%ED%95%B4-%ED%8C%8C%EC%9D%BC-Dto-%EB%8F%99%EC%8B%9C-%EC%9A%94%EC%B2%AD-%EC%8B%9C-%EB%B0%9C%EC%83%9D-%EC%97%90%EB%9F%AC)

Unagi-zoso committed:
ROFO-42 MaxFileSizeValidator 작성

Unagi-zoso committed:
ROFO-42 커스텀 유효 검사 시 발생하는 예외 핸들링 로직 추가

Unagi-zoso committed:
ROFO-42 WebPValidator 타켓 VALUE_PARAMETER 추가

Unagi-zoso committed:
ROFO-42 SocialAccessToken 클래스 제거

당장 필요한 기능이 아니라 String 으로 바꿉니다.

Unagi-zoso committed:
ROFO-42 AuthController @RequestPart 적용 및 스웨거 어노테이션 추가

Unagi-zoso committed:
ROFO-42 AuthCommandService register 매개변수 추가

Unagi-zoso committed:
ROFO-42 중복 닉네임 검사 API 추가

Unagi-zoso committed:
ROFO-42 회원가입 시 닉네임 중복 여부 검사 추가

Unagi-zoso committed:
ROFO-42 DuplicatedNicknameResponse 변환 역할 서비스로 변경

기존엔 컨트롤러에서 다뤘는데 이런 로직은 서비스에서 담당하는게 좋을 것 같아 바꿨다.

Unagi-zoso commented:
저희 개발환경에서 1MB 이상의 파일을 넣어봤는데 다음 결과가 나왔습니다. 
좀 찾아보니 nginx 나 쿠버네티스 인그레스 설정을 고려해보라고 하네요.

<img width="543" alt="스크린샷 2024-06-25 오후 7 47 58" src="https://github.com/Weit-2nd/roady-foody-server/assets/97292432/af6ee318-c469-4b2f-9d2b-5c256f0ac7a0">

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/presentation/api/AuthControllerTest.kt:
```diff
@@ -0,0 +1,141 @@
+                    verify(exactly = 0) { authCommandService.register(any<String>(), any<SignUpRequest>(), any<MultipartFile>()) }
+                }
+            }
+
```
iamge가 1MB를 넘으면 어떻게 될지에 대한 테스트도 추가되면 좋을것 같습니다~!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/application/service/AuthCommandService.kt:
```diff
@@ -0,0 +1,59 @@
+    ) {
+        val socialId = obtainSocialId(signUpRequest.socialLoginType, socialAccessToken)
+
+        if (userRepository.existsBySocialId(socialId)) {
+            log.error("UserAlreadyExistsException={}", socialId)
+            throw UserAlreadyExistsException()
+        }
+        if (userRepository.existsByProfileNickname(signUpRequest.nickname)) {
+            throw UserAlreadyExistsException()
+        }
```
요 두개 터트리는 에러가 같은데 if문을 합치는게 낫지 않을까요??

konempty commented on src/main/kotlin/kr/weit/roadyfoody/auth/presentation/spec/AuthControllerSpec.kt:
```diff
@@ -0,0 +1,150 @@
+        description = "회원가입 시 사용할 닉네임이 중복되는지 검사합니다.",
+        parameters = [
+            Parameter(name = "nickname", description = "닉네임", required = true, example = "테스트닉네임입니다"),
+        ],
```
요 API는 ApiResponse를 정의 안해줘도 클라분들이 안헷갈리실까요??

konempty commented on src/main/kotlin/kr/weit/roadyfoody/global/validator/MaxFileSizeValidator.kt:
```diff
@@ -0,0 +1,31 @@
+    val payload: Array<KClass<out Payload>> = [],
+)
+
+class MaxFileSizeValidator : ConstraintValidator<MaxFileSize, MultipartFile?> {
```
validator에 대한 테스트도 있으면 좋을것 같아요~!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/global/validator/WebPValidator.kt:
```diff
@@ -0,0 +1,25 @@
+    val payload: Array<KClass<out Payload>> = [],
+)
+
+class WebPImageValidator : ConstraintValidator<WebPImage, MultipartFile?> {
+    override fun isValid(
+        p0: MultipartFile?,
+        p1: ConstraintValidatorContext?,
+    ): Boolean = p0 == null || MimeUtils.detectMimeType(p0) == "image/webp"
+}
```
얘도 테스트가 있으면 좋을것 같구요 ㅎㅎ

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/fixture/AuthFixture.kt:
```diff
@@ -0,0 +1,47 @@
+
+const val TEST_IMAGE_FILE_NAME = "testImage"
+
+fun createTestImageFile(
+    format: ImageFormat,
+    name: String = TEST_IMAGE_FILE_NAME,
+    originalName: String = TEST_IMAGE_FILE_NAME,
+): MultipartFile =
+    MockMultipartFile(
+        name,
+        name,
+        "image/$format.value",
+        generateImageBytes(format),
+    )
+
+const val PROFILE_IMAGE_FILE_NAME = "profileImage"
+
+fun createProfileImageFile(
+    format: ImageFormat,
+    name: String = PROFILE_IMAGE_FILE_NAME,
+    originalName: String = PROFILE_IMAGE_FILE_NAME,
+): MockMultipartFile =
+    MockMultipartFile(
+        name,
+        originalName,
+        "image/$format.value",
+        generateImageBytes(format),
+    )
+
```
이미지 생성 로직이 다 동일한것 같은데 굳이 사용용도 별로 나눠둔 이유가 있을까요??

konempty commented on src/test/kotlin/kr/weit/roadyfoody/auth/fixture/AuthFixture.kt:
```diff
@@ -0,0 +1,47 @@
+
+const val SIGN_UP_REQUEST_FILE_NAME = "signUpRequest"
+
+fun createSignUpRequestFile(
+    contentStream: InputStream,
+    name: String = SIGN_UP_REQUEST_FILE_NAME,
+    originalName: String = SIGN_UP_REQUEST_FILE_NAME,
+    contentType: String = MediaType.APPLICATION_JSON_VALUE,
+): MockMultipartFile = MockMultipartFile(name, originalName, contentType, contentStream)
```
얘도 쓰임새가 많아보이는데 좀더 보편적으로 사용가능한 이름으로 지어서 다른곳에서 재활용하기 쉽도록 하면 어떨까요??

konempty commented on src/test/kotlin/kr/weit/roadyfoody/support/annotation/ControllerIntegrateTest.kt:
```diff
@@ -0,0 +1,19 @@
+@Transactional
+@Testcontainers
+@ContextConfiguration(initializers = [TestContainersConfig.Initializer::class])
+annotation class ControllerIntegrateTest
```
얘는 결국 쓰임새가 없어진것 같은데 없애면 어떨까요?? ㅋㅋㅋ

konempty reviewed (approved):
작업기간보다 리뷰 반영 기간이 더 길었던것 같네요 ㅋㅋㅋ큐ㅠㅠㅠ
앞으로 얼라인이 되어가면서 리뷰는 조금씩 줄지않을까 싶네요
고생하셨구요 코멘트 조금 남겼는데 적용하신뒤 머지 부탁드립니다~!

konempty commented:
>최대 파일 크기를 검사하는 기능을 추가했습니다.
이 기능을 코드로 테스트하는 건 생략했습니다.
기준을 1MB 로 했는데 동적으로 1MB 의 파일을 생성하는건 이미지 포매팅 과정에서 압축 등을 고려했을 때
정확하게 만들기 힘들고 생성 비용이 너무 컸습니다.
그래서 개인환경에서 직접 1MB 크기 이상의 파일을 넣어봄으로써 테스트했습니다.
혹시 좋은 방법이 있을까요?
저희 개발환경에서 1MB 이상의 파일을 넣어봤는데 다음 결과가 나왔습니다.
좀 찾아보니 nginx 나 쿠버네티스 인그레스 설정을 고려해보라고 하네요.

어후 ㅋㅋㅋ 인그레스가 막고있나보군요
저건 수정해두겠습니다~!

konempty commented:
@Unagi-zoso 인그레스 수정완료했습니다~!

konempty commented on src/main/kotlin/kr/weit/roadyfoody/global/client/RestClientConfig.kt:
```diff
@@ -25,6 +26,7 @@ class RestClientConfig {
         private const val CONNECT_TIME = 1L
         private const val READ_TIME = 5L
         private const val TOURISM_URL = "http://apis.data.go.kr/B551011/KorService1"
+        private const val KAKAO_URL = "https://kapi.kakao.com"
```
요거 지금알았는데 카카오 지도 검색도있어서 카카오 URL이라 쓰면 곤란할것 같아요 ㅋㅋㅋ
좀더 직관적인 이름으로 부탁드려요!
